src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,621 @@
+                prologue = previousPrologue;
+                return;
+            }
+
+            Debug.Assert(_factory.TopLevelMethod is not null);
+            Debug.Assert(_factory.CurrentFunction is not null);
+
+            var isStateMachine = _factory.CurrentFunction.IsAsync || _factory.CurrentFunction.IsIterator;
+
+            var prologueBuilder = ArrayBuilder<BoundStatement>.GetInstance(_factory.CurrentFunction.ParameterCount);
+
+            foreach (var parameter in _factory.CurrentFunction.Parameters)
+            {
+                if (parameter.RefKind == RefKind.Out || parameter.IsDiscard)
+                {
+                    continue;
+                }
+
+                var parameterLogger = GetLocalOrParameterStoreLogger(parameter.Type, parameter, refAssignmentSourceIsLocal: null, _factory.Syntax);
+                if (parameterLogger != null)
+                {
+                    prologueBuilder.Add(_factory.ExpressionStatement(_factory.Call(receiver: _factory.Local(_scope.ContextVariable), parameterLogger,
+                        MakeStoreLoggerArguments(parameterLogger.Parameters[0], parameter, parameter.Type, _factory.Parameter(parameter), refAssignmentSourceIndex: null, _factory.Literal((ushort)parameter.Ordinal)))));
+                }
+            }
+
+            if (previousPrologue != null)
+            {
+                prologueBuilder.Add(previousPrologue);
+            }
+
+            prologue = _factory.StatementList(prologueBuilder.ToImmutableAndFree());
+
+            var (entryOverload, entryArgs) = (isLambdaBody, isStateMachine) switch
+            {
+                (isLambdaBody: false, isStateMachine: false) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogMethodEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod) }),
+                (isLambdaBody: true, isStateMachine: false) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogLambdaEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.MethodDefIndex(_factory.CurrentFunction) }),
+                (isLambdaBody: false, isStateMachine: true) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogStateMachineMethodEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.StateMachineInstanceId() }),
+                (isLambdaBody: true, isStateMachine: true) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogStateMachineLambdaEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.MethodDefIndex(_factory.CurrentFunction), _factory.StateMachineInstanceId() }),
+            };
+
+            var entryLogger = GetWellKnownMethodSymbol(entryOverload, _factory.Syntax);
+            var instrumentationPrologue = (entryLogger != null) ?
```

Comment:
>entryLogger != null [](http://example.com/codeflow?start=43&length=19)

Are there non-error scenarios where the well-known methods are missing?

Reply:
Only error scenarios. Added a test that validates emit diagnostics in such cases.